### ISSUE TITLE
NBXplorer is stuck at sync

### ISSUE THREAD

**Issue by MarcosMe** [btcpayserver/btcpayserver-docker#185]
My btcpayserver doens't fully start as nbxplorer is stuck at sync:

docker logs --tail 10 generated_nbxplorer_1
info: Explorer:       BTC: Handshaked
info: Configuration:  BTC: Loading chain from node
info: Explorer:       BTC: Loading chain...
info: Explorer:       BTC: Chain loaded
info: Explorer:       BTC: NBXplorer is correctly whitelisted by the node
info: Configuration:  BTC: Height: 591833
info: Configuration:  BTC: Saving chain to cache...
info: Configuration:  BTC: Chain cached
info: Explorer:       BTC: Starting scan at block 578637
info: Events:         BTC: Node state changed: NotStarted => NBXplorerSynching

Even ./btcpay-down.sh / ./btcpay-up.sh doesn't solve the problem.
Is there a way to force or restart only nbxplorer?

**Comment by MarcosMe** [comment]
![BTC](https://user-images.githubusercontent.com/28023775/63700307-8bb34f80-c81a-11e9-853b-41a4fe92ec47.jpg)

**Comment by NicolasDorier** [comment]
did you stopped BTCpay for long time? What can happen is that you stopped BTCpay for several month, and when you restarted, bitcoin started pruning blocks before NBXplorer could synchronize.

**Comment by NicolasDorier** [comment]
There is several way I can fix your instance, but I need to know if you have anything important on your server, when did you created this server, and when did you restarted it.

**Comment by MarcosMe** [comment]
> did you stopped BTCpay for long time? 
No, I've been trying it for some time, and I've done a few times this WA (cause 1):
[https://docs.btcpayserver.org/faq-and-common-issues/faq-synchronization](url)
And afterwards I've done the fast sync part:
[https://docs.btcpayserver.org/deployment/raspberrypideployment](url)
It was created long time ago, but restarted recently (it had a problem with external HD).

**Comment by NicolasDorier** [comment]
So if you have nothing important in your instance 

```
btcpay-down.sh
docker volume rm -f generated_nbxplorer_datadir
btcpay-up.sh
```

**Comment by MarcosMe** [comment]
Hello Nicolas, 
Thank you for the fast feedback.
After a while the sync started to show in nbxplorer logs, and I though everything was ok, just had to wait for full sync.
But then it got stuck at at some block (it is like this for quite some time, and it happened the same in the past):

docker logs --tail 10 generated_nbxplorer_1
info: Events:         BTC: New block 000000000000000000158092ed4c6d400aad9270d3b7bf49d99fe1a26bab9ba9 (591891)
info: Events:         BTC: New block 00000000000000000006fd28b51ad8126064ba07930db5c2051614d9fe2b3cb8 (591892)
info: Events:         BTC: New block 0000000000000000000ba61f79d4bd0d074332f84bea38fbcd88225e579394c8 (591894)
info: Events:         BTC: New block 00000000000000000001b61213f2d508d19f9799cdeb8b0bac54f5d1ce788e79 (591893)
info: Events:         BTC: New block 0000000000000000001228307b11a38f14eded593b781ce44304774dfd962d10 (591895)
info: Events:         BTC: New block 0000000000000000000e9c1c75b7875df8c77d061ccbe443ec4783a531d8d1f2 (591897)
info: Events:         BTC: New block 000000000000000000178838f52662dcc871a3bc511a115b26969b0e62da532c (591896)
info: Events:         BTC: New block 00000000000000000006879bc0d004c6ea49201647023c5823bc5854303bf5a3 (591898)
info: Explorer:       BTC: Block download seems to hang, let's reconnect to this node
info: Explorer:       BTC: Closed connection with node

BTC

    NBXplorer headers height: 591898
    The node is synchronized (Height: 591958)

Do you still think the best option is to remove generated_nbxplorer_datadir?

**Comment by NicolasDorier** [comment]
@MarcosMe just do a `docker restart generated_nbxplorer_1` . This is a different problem, likely a bug. I need to fix that.

**Comment by MarcosMe** [comment]
Hello @NicolasDorier 
After docker restart generated_nbxplorer_1 problem happened again, in a different block:
docker logs --tail 10 generated_nbxplorer_1
info: Events:         BTC: New block 00000000000000000011bf5d6b04d802c1ed5a6e7e25dc460b64c88a236d7899 (592040)
info: Events:         BTC: New block 00000000000000000006eb57a3031c2558c35188428d8ee71592dc39977152d3 (592039)
info: Events:         BTC: New block 0000000000000000001722cd27e9804a474d59af967ad2df06f4f9bb425fa7e7 (592043)
info: Events:         BTC: New block 0000000000000000000dcb7912c3ce17151a13d35d2deeb6c3036e80865dc012 (592038)
info: Events:         BTC: New block 00000000000000000001ca5a636e336716019a4126c568f3e2e01399cff59fe3 (592041)
info: Events:         BTC: New block 0000000000000000001b853668b78722157a53560078d9ac38e7b1b74d5700c5 (592042)
info: Events:         BTC: New block 0000000000000000000b32069d7e7ba341e0890f41dac8ab4c9037b716029838 (592045)
info: Events:         BTC: New block 0000000000000000000884a77f5ff65c397cae88fb385e10cf5df751a088e6c9 (592046)
info: Explorer:       BTC: Block download seems to hang, let's reconnect to this node
info: Explorer:       BTC: Closed connection with node

And in btcpayserver web page:
Your nodes are synching...
Your node is synching the entire blockchain and validating the consensus rules... 
BTC
    NBXplorer headers height: 592046
    The node is synchronized (Height: 592109)

Let me know if you need any other information, and any action I can perform. Should I try restart again?

**Comment by NicolasDorier** [comment]
@MarcosMe try again. This is strange it happen so often, what environment are you running nbxplorer on?

**Comment by NicolasDorier** [comment]
@MarcosMe I think I found the bug upstream that block NBXplorer. That said, it is strange it happen so often to you.

**Comment by MarcosMe** [comment]
@NicolasDorier I tried again the restart. I will let you know the outcome.
Concerning the environment, I'm using a raspberry pi 3 ARMv7l (with raspian stretch), with an external HD (WD PiDrive). And with docker as per raspberry deployment:
https://docs.btcpayserver.org/deployment/raspberrypideployment

**Comment by NicolasDorier** [comment]
@MarcosMe I see. Don't forget to use `opt-save-memory` on rapberry 3. I will probably release new version of nbxplorer tomorrow. I found out why it would not reconnect automatically.

**Comment by MarcosMe** [comment]
@NicolasDorier Thank you! I confirm `opt-save-memory` is used.

**Comment by NicolasDorier** [comment]
@MarcosMe I just pushed an update, if you run `btcpay-update.sh` it should update NBXPlorer and you should not have problem anymore. (if is stall during sync, it should reconnect)

**Comment by MarcosMe** [comment]
Hello @NicolasDorier 
Unfortunately still not working, now it doesn't stall at some block, but it seems to restart once it ends the sync (behaved like this already twice):

info: Events:         BTC: New block 000000000000000000168c7ff5a2618d4aade9f92915090cef08017092b09ea5 (592399)
info: Events:         BTC: New block 000000000000000000078374c458b9ebb20de4e770abe3fdfa856eeec2941a89 (592398)
info: Events:         BTC: New block 0000000000000000000a8cdc8c5a4ca820603d9a188f8749b72c3c838cf22697 (592400)
info: Events:         BTC: New block 0000000000000000001304f663d150fa6d20cbefd88ec5ef8f593ee4d6a53754 (592406)
info: Events:         BTC: New block 0000000000000000001559b880c34e0d34ce76a55e1a3f3b8f83897584fb2b94 (592401)
info: Events:         BTC: New block 0000000000000000001850c3f937996b1b14e69e8e903ef21c4fb6e64829d81e (592404)
info: Events:         BTC: New block 0000000000000000000a7f06105cae52513ccffedcb017e9fd5458863b3c79e9 (592407)
info: Events:         BTC: xpub6BertP...TTixdQAX34fhD-[p2sh] matching confirmed transaction 10dbd6...d34eaf (0/1)
info: Events:         BTC: New block 000000000000000000059523802a51ccbf5464df1fcc2a4d7496e1512b3c8dfb (592405)
info: Events:         BTC: New block 000000000000000000123bb3212d8e48a9c0a9e654b7b6043e3e41ad3c22951f (592411)
info: Events:         BTC: New block 00000000000000000017d59a91344f63dc77a5c92bf12c69a408c1590da4c981 (592410)
info: Events:         BTC: New block 00000000000000000017c7d369327a7a3243040c27bd5e537042c743ec70f431 (592409)
info: Events:         BTC: New block 000000000000000000023f0762797d34e5acfe4d3d70d8b174ebdc1c7ed83c17 (592408)
info: Events:         BTC: New block 000000000000000000009493182cbf9b6b51a55ed095dfe2324549b896de154f (592412)
info: Events:         BTC: New block 0000000000000000000e52610668441f3c161e3f5059df68d1a12a4ed413ca19 (592415)
info: Events:         BTC: New block 0000000000000000000501729aa61c17afd54030673636f579c6db1d029048f4 (592414)
info: Events:         BTC: New block 0000000000000000001972895549f4004a465a75bf2a5a947314a0770e471455 (592413)
info: Events:         BTC: New block 0000000000000000000b09f1a1945b13f284d16b8b8f028564d700b62353bd57 (592417)
info: Events:         BTC: New block 000000000000000000050ec94592918fd112dee63baba4f04c9db10a7fe9ff7a (592416)
info: Events:         BTC: New block 00000000000000000002404ba39196a3feecae3f7d2ef1e47ae7d3bad7642da3 (592420)
info: Events:         BTC: New block 0000000000000000000e93fbecdd3640ebc5a6c2cbe170bbeaca7332fe1c7bf8 (592421)
info: Events:         BTC: New block 0000000000000000000c2443f7992789c66dc4aa4c14bbcb2fb6a7fcd13b298b (592419)
info: Events:         BTC: New block 00000000000000000006a703e5629b4446052fafcee4c37719e6f93a2258e730 (592422)
info: Events:         BTC: New block 0000000000000000000caeb8e99e6911be88699375418a5affbbeb2585cb0090 (592423)
info: Events:         BTC: New block 000000000000000000019f6b5756eb705b40a8b0302d0cbdc614e1273a24676d (592418)
info: Events:         BTC: New block 000000000000000000038c0af2b17785ebcbca6fde7af95f90a65c116f5316b8 (592424)
info: Explorer:       BTC: Block download seems to hang, let's reconnect to this node
info: Events:         BTC: Node state changed: NBXplorerSynching => NotStarted
info: Configuration:  BTC: Testing RPC connection to http://bitcoind:43782/
info: Explorer:       BTC: Closed connection with node
info: Configuration:  BTC: RPC connection successful
info: Configuration:  BTC: Full node version detected: 180000
info: Configuration:  BTC: Loading chain from cache...
info: Configuration:  BTC: Height: 592352
info: Configuration:  BTC: Trying to connect via the P2P protocol to trusted node (bitcoind:39388)...
info: Explorer:       BTC: TCP Connection succeed, handshaking...
info: Explorer:       BTC: Handshaked
info: Configuration:  BTC: Loading chain from node
info: Explorer:       BTC: Loading chain...
info: Explorer:       BTC: Chain loaded
info: Explorer:       BTC: NBXplorer is correctly whitelisted by the node
info: Configuration:  BTC: Height: 592424
info: Configuration:  BTC: Saving chain to cache...
info: Configuration:  BTC: Chain cached
info: Explorer:       BTC: Starting scan at block 578637
info: Events:         BTC: Node state changed: NotStarted => NBXplorerSynching

**Comment by NicolasDorier** [comment]
@MarcosMe wtf `Starting scan at block 578637` it should not rescan at 578637. You are using the normal docker deploy right?

**Comment by MarcosMe** [comment]
@NicolasDorier Yes, I'm using normal docker deployment, I've seen this number in the past, but not sure why it appears so often. No haven't modified NBXplorer config.
The only thing I can recall of doing (differently) due to HD problems was the WA (cause 1) from sync issues:
https://docs.btcpayserver.org/faq-and-common-issues/faq-synchronization
And afterwards I've done the fast sync part:
https://docs.btcpayserver.org/deployment/raspberrypideployment

**Comment by NicolasDorier** [comment]
@MarcosMe I will add additional logs to pin point what happen will make new version in 30min

**Comment by MarcosMe** [comment]
@NicolasDorier Many thanks. Let me know once it is ready to update

**Comment by NicolasDorier** [comment]
@MarcosMe can you update `$BTCPAY_DOCKER_COMPOSE` with VIM and replace the version of NBXPlorer from 2.0.0.57 to 2.0.0.58 manually ?

Secondly, in the same file, next to where you change the version, for nbxplorer, add this like with the same indentation as others:

```
NBXPLORER_BTCSTARTHEIGHT: 592000
```

Then `btcpay-up.sh` to make sure it takes this version.

**Comment by MarcosMe** [comment]
Hello @NicolasDorier 
It worked, btcpayserver is fully synched :)
Thanks a lot for your support on this.

**Comment by NicolasDorier** [comment]
Run `btcpay-update.sh` to remove the changes, else it will rescan at 592000 whne it restarts.

**Comment by MarcosMe** [comment]
Yes, I've already done it, and restarted btcpayserver successfully :)
